pep-0xxx.rst

```diff
@@ -0,0 +1,1251 @@
+directly, instead preferring to create new contexts from new TLSConfiguration
+objects. Obviously, the ABCs cannot enforce this constraint, and so they do not
+attempt to.
+
+The ``Context`` abstract base class has the following class definition::
+
+    TLSBufferObject = Union[TLSWrappedSocket, TLSWrappedBuffer]
+
+
+    class _BaseContext(metaclass=ABCMeta):
+        @abstractmethod
+        def __init__(self, configuration: TLSConfiguration):
+            """
+            Create a new context object from a given TLS configuration.
+            """
+
+        @property
+        @abstractmethod
+        def configuration(self) -> TLSConfiguration:
+            """
+            Returns the TLS configuration that was used to create the context.
+            """
+
+
+    class ClientContext(_BaseContext):
+        @abstractmethod
+        def wrap_socket(self,
+                        socket: socket.socket,
+                        server_hostname: Optional[str],
+                        auto_handshake=True: bool) -> TLSWrappedSocket:
+            """
+            Wrap an existing Python socket object ``socket`` and return a
+            ``TLSWrappedSocket`` object. ``socket`` must be a ``SOCK_STREAM``
+            socket: all other socket types are unsupported.
+
+            The returned SSL socket is tied to the context, its settings and
+            certificates.
+
+            The parameter ``server_hostname`` specifies the hostname of the
+            service which we are connecting to. This allows a single server to
+            host multiple SSL-based services with distinct certificates, quite
+            similarly to HTTP virtual hosts. This is also used to validate the
+            TLS certificate for the given hostname. If hostname validation is
+            not desired, then pass ``None`` for this parameter.
+
+            The parameter ``auto_handshake`` specifies whether to do the SSL
+            handshake automatically after doing a ``socket.connect()``, or
+            whether the application program will call it explicitly, by
+            invoking the ``TLSWrappedSocket.do_handshake()`` method. Calling
+            ``TLSWrappedSocket.do_handshake()`` explicitly gives the program
+            control over the blocking behavior of the socket I/O involved in
+            the handshake.
+            """
+
+        @abstractmethod
+        def wrap_buffers(self, incoming: Any, outgoing: Any,
+                         server_hostname: Optional[str]) -> TLSWrappedBuffer:
+            """
+            Wrap a pair of buffer objects (``incoming`` and ``outgoing``) to
+            create an in-memory stream for TLS. The SSL routines will read data
+            from ``incoming`` and decrypt it, and write encrypted data to
+            ``outgoing``.
+
+            The buffer objects must be either file objects or objects that
+            implement the buffer protocol.
```

Comment:
Good question. I don't know that it can, tbh, though it bears further investigation.